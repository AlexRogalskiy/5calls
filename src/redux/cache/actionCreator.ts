import { Group, CacheableGroup } from './../../common/model';
import { CacheGroupAction, AppCacheActionType } from './';

export const cacheGroupActionCreator = (cgroup: CacheableGroup): CacheGroupAction => {
  return {
    type: AppCacheActionType.CACHE_GROUP,
    payload: cgroup
  };
};

export const addGroupToCache = (group: Group) => {
  return {
    type: AppCacheActionType.ADD_GROUP_TO_CACHE,
    payload: group
  };
};
