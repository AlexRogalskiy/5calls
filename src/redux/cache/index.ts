export { AppCache, addOrReplaceCacheableGroup, findCacheableGroup, CacheTimeoutData } from './cache';
export { AppCacheAction, AppCacheActionType, CacheGroupAction, AddToCacheAction } from './action';
export { cacheGroup, addToCache } from './asyncActionCreator';
export { appCacheReducer } from './reducer';