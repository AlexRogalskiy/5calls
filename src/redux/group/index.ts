export {
  GroupActionType,
  GroupLoadingStatusAction,
  GroupLoadingActionStatus,
  CurrentGroupAction } from './action';
export {
  setDefaultGroupLoadingAction,
  setFoundGroupLoadingAction,
  setNotFoundGroupLoadingAction,
  setErrorGroupLoadingAction,
  setCurrentGroupAction
} from './actionCreator';
export { GroupState, groupStateReducer } from './reducer';
export { updateGroup } from './asyncActionCreator';
