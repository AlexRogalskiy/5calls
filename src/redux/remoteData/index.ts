export { IssuesAction, RemoteDataAction, RemoteDataActionType,
  CallCountAction, ApiErrorAction, DonationsAction } from './action';
export { issuesActionCreator, callCountActionCreator,
  apiErrorMessageActionCreator, donationsActionCreator } from './actionCreator';
export { RemoteDataState, remoteDataReducer } from './reducer';
export { startup, getIssuesIfNeeded,
   fetchCallCount, fetchBrowserGeolocation } from './asyncActionCreator';
