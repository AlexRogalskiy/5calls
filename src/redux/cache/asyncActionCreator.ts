import { Dispatch } from 'react-redux';
import { ApplicationState } from '../root';
import { cacheGroupActionCreator, addGroupToCache } from './actionCreator';
import {
  hasGroupCacheTimeoutExceeded,
  findCacheableGroup, cacheableGroupFactory } from './cache';
import * as api from '../../services/apiServices';
import { Group, CacheableGroup } from '../../common/model';

export const cacheGroup = (groupId: string) => {
  return (
    dispatch: Dispatch<ApplicationState>,
    getState: () => ApplicationState
  ): Promise<void> => {
      const state = getState();
      const appCache = state.appCache;
      let cacheable: CacheableGroup | undefined = findCacheableGroup(groupId, appCache);
      // Call API to get group data if group is not found
      // in cache or cache timeout has exceeded.
      if ( !cacheable ||
        hasGroupCacheTimeoutExceeded({ timestamp: cacheable.timestamp })) {
        return api.getGroup(groupId)
          .then(group => {
            const cacheableGroup = cacheableGroupFactory(group);
            if (cacheableGroup) {
              dispatch(cacheGroupActionCreator(cacheableGroup));
            }
            return Promise.resolve();
          });
      } else {
        // cacheable group is already in cache
        return Promise.resolve();
      }

    };
};

export const addToCache = (group: Group) => {
  return (
    dispatch: Dispatch<ApplicationState>,
    getState: () => ApplicationState): Promise<void> => {
      dispatch(addGroupToCache(group));
      return Promise.resolve();
    };
};
