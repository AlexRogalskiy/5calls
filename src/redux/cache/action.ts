import { Group, CacheableGroup } from './../../common/model';
import { Action } from 'redux';

export enum AppCacheActionType {
  CACHE_GROUP = 'CACHE_GROUP',
  ADD_GROUP_TO_CACHE = 'ADD_GROUP_TO_CACHE'
}

export interface AppCacheAction extends Action {
  type: AppCacheActionType;
  payload?: {};
}

export interface CacheGroupAction extends AppCacheAction {
  type: AppCacheActionType.CACHE_GROUP;
  payload: CacheableGroup;
}

export interface AddToCacheAction extends AppCacheAction {
  type: AppCacheActionType.ADD_GROUP_TO_CACHE;
  payload: Group;
}
