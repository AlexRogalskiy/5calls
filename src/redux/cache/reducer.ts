import { Reducer } from 'redux';
import {
  AppCache, AppCacheAction, AppCacheActionType,
  addOrReplaceCacheableGroup } from './';
import { Group, CacheableGroup } from '../../common/model';

export const appCacheReducer: Reducer<AppCache> = (
  state: AppCache = {groups: []} as AppCache,
  action: AppCacheAction): AppCache => {

    switch (action.type) {
      case AppCacheActionType.CACHE_GROUP:
        const groups = state.groups;
        const cgroup = action.payload as CacheableGroup;
        const group = cgroup.group;
        const newGroups = addOrReplaceCacheableGroup(groups, group);
        return { ...state, groups: newGroups };
      case AppCacheActionType.ADD_GROUP_TO_CACHE:
        const groups1 = state.groups;
        const group1 = action.payload as Group;
        const newGroups1 = addOrReplaceCacheableGroup(groups1, group1);
        return { ...state, groups: newGroups1 };
      default:
        return state;
    }
};
