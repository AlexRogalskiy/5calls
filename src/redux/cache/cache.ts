import { CacheTimeoutData } from './cache';
import { Group, CacheableGroup } from '../../common/model';
import { cacheTimeout } from './../../common/constants';

export type GroupCache = CacheableGroup[];

/**
 * The cache for this application with fields
 * containing the data to be cached.
 *
 */
export class AppCache {
  groups: GroupCache;
  constructor(groupCache: GroupCache) {
      this.groups = groupCache;
  }
}

/**
 * Finds the Group and timestamp in the cache.
 *
 * @param {string} id the group's id
 * @param {AppCache} cache the cache stored in the redux store
 */
export function findCacheableGroup(id: string, cache: AppCache): CacheableGroup | undefined {
    return cache.groups ? cache.groups
        .filter(g => g.group.id === id)
        .pop() : undefined;
    }

/**
 * Factory for `CacheableGroup` objects.
 *
 * @export
 * @param {Group} group - group to be cached
 * @returns {CacheableGroup | undefined}
 */
export function cacheableGroupFactory(group: Group): CacheableGroup | undefined {
    const now = new Date().getTime();
    if (group) {
      return { group, timestamp: now };
    } else {
      return undefined;
    }
}

export function addOrReplaceCacheableGroup(groups: CacheableGroup[], group: Group): CacheableGroup[] {
    // remove old group if present
    const newGroups = groups.filter(cgroup => {
        return cgroup.group.id !== group.id;
    });
    const newcgroup = cacheableGroupFactory(group);
    if (newGroups && newcgroup) {
        return [...newGroups, newcgroup];
    } else {
        return groups;
    }
}

/**
 * Holds data to determine if cached data has timed out.
 * The `timestamp` field is the only required field
 * as it is the timestamp that is used to determine if
 * a timeout has exceeded.
 * The `now` field represents the current timestamp, which
 * by default should be the value of new Date().getTime()
 * The `timeout` field is the timeout value (in milliseconds),
 * which should be a constant set in `constants.ts`. Both the
 * now and timeout arguments should only be used for testing.
 *
 * This interface was created to make sure hasTimeoutExceeded()
 * arguments did not get mixed up since all of them are of
 * number type.
 */
export interface CacheTimeoutData {
    timestamp: number;
    now?: number; // current timestamp
    timeout?: number; // timeout value in milliseconds
}

/**
 * Determines if a cache timeout has exceeded.
 *
 * By default, the timeout value comes from
 * `cacheTimeout.default` in `constants.ts`
 *
 * @param {CacheTimeoutData} timeoutData
 * @returns { boolean } - true if timeout has exceeded;
 * otherwise false.
 */
export function hasCacheTimeoutExceeded(timeoutData: CacheTimeoutData): boolean {
    const now = timeoutData.now || new Date().getTime();
    const timeout = timeoutData.timeout || cacheTimeout.default;
    return now - timeoutData.timestamp > timeout;
}

/**
 * Determines if a cache timeout has exceeded for group/team
 * data.
 *
 * By default, the timeout value comes from
 * `cacheTimeout.groups` in `constants.ts`
 *
 * @param {CacheTimeoutData} timeoutData
 * @returns { boolean } - true if timeout has exceeded;
 * otherwise false.
 */
export function hasGroupCacheTimeoutExceeded(timeoutData: CacheTimeoutData): boolean {
    const timestamp = timeoutData.timestamp;
    const timeout = cacheTimeout.groups;
    return hasCacheTimeoutExceeded({timestamp, timeout});
}