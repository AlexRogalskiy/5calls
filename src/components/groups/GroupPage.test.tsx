import * as React from 'react';
import { shallow } from 'enzyme';
import { GroupPage } from './';
import { AppCache } from '../../redux/cache';
import { GroupState } from '../../redux/group';
import { RemoteDataState } from '../../redux/remoteData';

test('Test Group Page loading', () => {
  const pageProps = initPage();
  const component = shallow(
    <GroupPage
      {...pageProps}
    />
  );
  expect(component).toMatchSnapshot();
});

const initPage = () => {
  const id = 'craig';
  return {
    match: {params: {groupid: id, issueid: '100'}, isExact: true, path: '', url: ''},
    appCache: AppCache,
    remoteState: {} as RemoteDataState,
    groupState: {} as GroupState,
  };
};
