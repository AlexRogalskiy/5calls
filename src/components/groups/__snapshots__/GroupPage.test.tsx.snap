// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Group Page loading 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
